### PR TITLE
mpl/ze: Fix initialization bugs

### PR DESCRIPTION
**Pull request by raffenet**
## Pull Request Description

Fix some bugs in the MPL ZE initialization code.
1. Make sure to get the correct number of devices.
2. Fix a potential memory leak if a GPU device is not found associated with a driver.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by raffenet** [comment]
test:mpich/ch4/most

File: src/mpl/src/gpu/mpl_gpu_ze.c [hunk 1]
```diff
@@ -93,6 +94,9 @@ static int gpu_ze_init_driver()
 
         if (NULL != global_ze_driver_handle) {
             break;
+        } else {
+            MPL_free(global_ze_devices_handle);
+            global_ze_devices_handle = NULL;
```
Comment (hzhou) [inline src/mpl/src/gpu/mpl_gpu_ze.c:99]: In the `else-` branch, `global_ze_devices_handle == NULL` already, right? What did I miss?

Comment (hzhou) [reply src/mpl/src/gpu/mpl_gpu_ze.c:99]: 🤦 I see.

**Comment by raffenet** [comment]
@zhenggb72 do you want to review these changes?

**Comment by zhenggb72** [comment]
Changes look good to me.

**Review by zhenggb72 — APPROVED** [review]

**Pull request merged**